### PR TITLE
CB-21892 Identify and exclude LOST nodes on YARN during fast upscale …

### PR DESCRIPTION
**Pull request by aakashshankar**
…and downscale

- Used CM API to filter nodes running NodeManager roles, with BAD or CONCERNING NODE_MANAGER_CONNECTIVITY healthCheck param
- Added said nodes to excludeList, which is removed from recommission and decommission list during fast upscale and downscale respectively

See detailed description in the commit message.

### CODE REVIEW

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 1]
```diff
@@ -54,10 +65,17 @@ public class ClouderaManagerClusterHealthService implements ClusterHealthService
...
+
+    static final String NODEMANAGER = "NODEMANAGER";
+
     private static final Logger LOGGER = LoggerFactory.getLogger(ClouderaManagerClusterHealthService.class);
 
     private static final Set<ApiHealthSummary> IGNORED_HEALTH_SUMMARIES = Sets.immutableEnumSet(
-            ApiHealthSummary.DISABLED,
+            DISABLED,
             ApiHealthSummary.NOT_AVAILABLE,
             ApiHealthSummary.HISTORY_NOT_AVAILABLE
```
Comment (adarshsharma1983) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:80]: You can static import NOT_AVAILABLE and HISTORY_NOT_AVAILABLE

File: common/src/main/resources/messages/messages.properties [hunk 2]
```diff
@@ -138,6 +139,7 @@ cluster.scaling.stopstart.downscale.starting.identifiedrecoverycandidates=Scalin
 cluster.scaling.stopstart.downscale.couldnotdecommission=Could not decommission services on all nodes. Some nodes may need attention: instanceCount={0}, hostname(s): [{1}]
 cluster.scaling.stopstart.downscale.enteringcmmaintmode=Putting {0} cm-host(s) into maintenance mode
 cluster.scaling.stopstart.downscale.enteredcmmaintmode=Finished putting {0} cm-host(s) into maintenance mode
+cluster.scaling.stopstart.downscale.exclude.lost.nodes=Found {0} disconnected host(s): [{1}]. Excluding them from the scaling operation.
```
Comment (adarshsharma1983) [inline common/src/main/resources/messages/messages.properties:142]: should we say scaling or downscaling?

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 3]
```diff
@@ -165,6 +201,25 @@ private static Optional<DetailedServicesHealthCheck> getDetailedServicesHealthCh
...
 
+    private Set<String> collectDisconnectedNodeManagers(List<ApiRole> allNMRoles) {
+        return allNMRoles.stream()
+                .filter(this::isNodeManagerDisconnected)
+                .filter(nm -> nonNull(nm.getHostRef()))
+                .map(nm -> nm.getHostRef().getHostname())
+                .collect(Collectors.toUnmodifiableSet());
+    }
+
+    private boolean isNodeManagerDisconnected(ApiRole role) {
```
Comment (sidseth) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:212]: How does this API behave for nodes which are in STOPPED state? Or have just been started - but the NodeManager is not yet in RUNNING state ?

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 4]
```diff
@@ -149,13 +185,13 @@ private static Optional<DetailedHostHealthCheck> getDetailedHostHealthCheck(ApiH
     private static Optional<DetailedServicesHealthCheck> getDetailedServicesHealthCheck(ApiHost apiHost, Optional<String> runtimeVersion) {
         if (CMRepositoryVersionUtil.isCmServicesHealthCheckAllowed(runtimeVersion)) {
             Set<String> servicesWithBadHealth = emptyIfNull(apiHost.getRoleRefs()).stream()
-                    .filter(roleRef -> Objects.nonNull(roleRef.getHealthSummary()))
-                    .filter(roleRef -> HealthCheckResult.UNHEALTHY.equals(healthSummaryToHealthCheckResult(roleRef.getHealthSummary())))
+                    .filter(roleRef -> nonNull(roleRef.getHealthSummary()))
+                    .filter(roleRef -> UNHEALTHY.equals(healthSummaryToHealthCheckResult(roleRef.getHealthSummary())))
```
Comment (sidseth) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:189]: Nit: Unnecessary changes which aren't related to the patch..

File: common/src/main/resources/messages/messages.properties [hunk 5]
```diff
@@ -123,6 +123,7 @@ cluster.scaling.stopstart.upscale.nodes.notstarted=Some instances did not reach
 cluster.scaling.stopstart.upscale.inadequate.nodes=Could not find adequate nodes to upscale(start) in hostGroup: {0}. Additional nodes may need to be added to the cluster. DesiredCount={1}, UpscaleCount={2}
 cluster.scaling.stopstart.upscale.commissioning=Commissioning services. Host group: {0}, InstanceCount={1}, hostname(s): [{2}]
 cluster.scaling.stopstart.upscale.commissioning2=Commissioning services. Host group: {0}, startedInstanceCount={1}, hostname(s): [{2}] recoveryCandidatesCount={3}, hostname(s): [{4}]
+cluster.scaling.stopstart.upscale.exclude.lost.nodes=Found {0} disconnected host(s): [{1}]. Excluding them from the scaling operation.
```
Comment (adarshsharma1983) [inline common/src/main/resources/messages/messages.properties:126]: can we say upscaling instead of scaling?

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java [hunk 6]
```diff
@@ -173,6 +177,19 @@ protected Selectable doAccept(HandlerEvent<StopStartDownscaleDecommissionViaCMRe
...
+    private void excludeAndLogDisconnectedNMsForDownscale(Stack stack, Map<String, InstanceMetadataView> hostsToRemove,
+            StopStartDownscaleDecommissionViaCMRequest request) {
+        ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+        Set<String> disconnectedNMHostNames = clusterHealthService.getDisconnectedNodeManagers();
+        if (!disconnectedNMHostNames.isEmpty()) {
+            LOGGER.info("Found {} disconnected NodeManagers: {}, for hostgroup {}. Excluding them from de-commission list", disconnectedNMHostNames.size(),
+                    disconnectedNMHostNames, request.getHostGroupName());
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_EXCLUDE_LOST_NODES,
+                    String.valueOf(disconnectedNMHostNames.size()), String.join(", ", disconnectedNMHostNames));
+            hostsToRemove.entrySet().removeIf(e -> disconnectedNMHostNames.contains(e.getKey()));
```
Comment (adarshsharma1983) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java:189]: This check is case sensitive. please check.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleCommissionViaCMHandler.java [hunk 7]
```diff
@@ -153,4 +157,17 @@ protected Selectable doAccept(HandlerEvent<StopStartUpscaleCommissionViaCMReques
...
+    private void excludeAndLogDisconnectedNMsForUpscale(Stack stack, Map<String, InstanceMetaData> hostsToRecommission,
+            StopStartUpscaleCommissionViaCMRequest request) {
+        ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+        Set<String> disconnectedNMHostNames = clusterHealthService.getDisconnectedNodeManagers();
+        if (!disconnectedNMHostNames.isEmpty()) {
+            LOGGER.info("Found {} disconnected NodeManagers: {} in hostgroup: {}. Excluding them from recommission list", disconnectedNMHostNames.size(),
+                    disconnectedNMHostNames, request.getHostGroupName());
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_UPSCALE_EXCLUDE_LOST_NODES,
+                    String.valueOf(disconnectedNMHostNames.size()), String.join(", ", disconnectedNMHostNames));
+            hostsToRecommission.entrySet().removeIf(e -> disconnectedNMHostNames.contains(e.getKey()));
```
Comment (adarshsharma1983) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleCommissionViaCMHandler.java:170]: same as above

File: cluster-cm/src/test/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthServiceTest.java [hunk 8]
```diff
@@ -229,6 +255,31 @@ void testServicesNotRunningOnHosts() throws ApiException {
         assertThat(result.areServicesNotRunning(hostName("host-3"))).isFalse();
     }
 
+    @Test
```
Comment (adarshsharma1983) [inline cluster-cm/src/test/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthServiceTest.java:258]: We should another test case(s) to check that getDisconnectedNodeManagers should return empty list.

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 1 above)
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Done

File: common/src/main/resources/messages/messages.properties (hunk 2 above)
Comment (aakashshankar) [reply common/src/main/resources/messages/messages.properties]: Yes, that makes more sense. Updated these messages

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 3 above)
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:216]: As discussed, the test will be DISABLED once the roles are stopped. [Jira comment with additional details](https://jira.cloudera.com/browse/CB-21892?page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel&focusedCommentId=5994369#comment-5994369)

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java (hunk 6 above)
Comment (aakashshankar) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java]: Converted all hostnames to lowercase and ensure lowercase is preserved during comparison

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 9]
```diff
@@ -54,12 +68,19 @@ public class ClouderaManagerClusterHealthService implements ClusterHealthService
 
     static final String HOST_AGENT_CERTIFICATE_EXPIRY = "HOST_AGENT_CERTIFICATE_EXPIRY";
 
+    // TODO Do we extend this to include additional health check params?
```
Comment (sidseth) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:71]: Convert to a jira and add to the comment, or remove the comment.

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 10]
```diff
@@ -54,12 +68,19 @@ public class ClouderaManagerClusterHealthService implements ClusterHealthService
...
+    static final String NODE_MANAGER_CONNECTIVITY = "NODE_MANAGER_CONNECTIVITY";
+
+    static final String YARN = "yarn";
+
+    static final String NODEMANAGER = "NODEMANAGER";
```
Comment (sidseth) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:76]: Look this up from YarnRoles ?

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 11]
```diff
@@ -54,12 +68,19 @@ public class ClouderaManagerClusterHealthService implements ClusterHealthService
 
     static final String HOST_AGENT_CERTIFICATE_EXPIRY = "HOST_AGENT_CERTIFICATE_EXPIRY";
 
+    // TODO Do we extend this to include additional health check params?
+    static final String NODE_MANAGER_CONNECTIVITY = "NODE_MANAGER_CONNECTIVITY";
+
+    static final String YARN = "yarn";
```
Comment (sidseth) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:74]: Is this the serviceName? Shouldn't this instead be going to the blueprint, figuring out the RoleType=YARN, and pulling the serviceName from the blueprint ?
(Looks like YarnServiceClient in Periscope should do the same as well)

If there's a blueprint where the YARN role is something other than 'yarn' (In the blueprint) for some reason - this will likely break, as well as the check in Periscope.

File: cluster-cm/src/test/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthServiceTest.java [hunk 12]
```diff
@@ -31,14 +41,26 @@
 
 class ClouderaManagerClusterHealthServiceTest {
 
+    private static final String RESOURCEMANAGER = "RESOURCEMANAGER";
```
Comment (sidseth) [inline cluster-cm/src/test/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthServiceTest.java:44]: YarnRoles

**Review by sidseth — COMMENTED** [review]
Mostly looks good. Re-using the constants for the YARN Role Names, and looking these up from the blueprint (instead of assuming "yarn") are the main change requests.

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 11 above)
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java:74]: Would this refer to the service refName or the serviceType itself?

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 9 above)
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Updated

File: cluster-cm/src/test/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthServiceTest.java (hunk 12 above)
Comment (aakashshankar) [reply cluster-cm/src/test/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthServiceTest.java]: Updated

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 10 above)
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Done

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 11 above)
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Changed this to pull the `refName` from the blueprint

**Review by Vatsal23082000 — APPROVED** [review]
LGTM

**Review by sidseth — APPROVED** [review]
Looks good. Could you please file a jira for Periscope making the same fix related to extracting the actual yarn service name from the blueprint instead of assuming "yarn"

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java [hunk 13]
```diff
@@ -173,6 +177,19 @@ protected Selectable doAccept(HandlerEvent<StopStartDownscaleDecommissionViaCMRe
...
+    private void excludeAndLogDisconnectedNMsForDownscale(Stack stack, Map<String, InstanceMetadataView> hostsToRemove,
+            StopStartDownscaleDecommissionViaCMRequest request) {
+        ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+        Set<String> disconnectedNMHostNames = clusterHealthService.getDisconnectedNodeManagers();
+        if (!disconnectedNMHostNames.isEmpty()) {
+            LOGGER.info("Found {} disconnected NodeManagers: {}, for hostgroup {}. Excluding them from de-commission list", disconnectedNMHostNames.size(),
+                    disconnectedNMHostNames, request.getHostGroupName());
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_EXCLUDE_LOST_NODES,
+                    String.valueOf(disconnectedNMHostNames.size()), String.join(", ", disconnectedNMHostNames));
+            hostsToRemove.entrySet().removeIf(e -> disconnectedNMHostNames.contains(StringUtils.lowerCase(e.getKey())));
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java:189]: Out of curiosity, what benefit does `StringUtils.lowerCase()` have vs `String.toLowerCase()`?

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleCommissionViaCMHandler.java [hunk 14]
```diff
@@ -153,4 +158,17 @@ protected Selectable doAccept(HandlerEvent<StopStartUpscaleCommissionViaCMReques
...
+    private void excludeAndLogDisconnectedNMsForUpscale(Stack stack, Map<String, InstanceMetaData> hostsToRecommission,
+            StopStartUpscaleCommissionViaCMRequest request) {
+        ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+        Set<String> disconnectedNMHostNames = clusterHealthService.getDisconnectedNodeManagers();
+        if (!disconnectedNMHostNames.isEmpty()) {
+            LOGGER.info("Found {} disconnected NodeManagers: {} in hostgroup: {}. Excluding them from recommission list", disconnectedNMHostNames.size(),
+                    disconnectedNMHostNames, request.getHostGroupName());
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_UPSCALE_EXCLUDE_LOST_NODES,
+                    String.valueOf(disconnectedNMHostNames.size()), String.join(", ", disconnectedNMHostNames));
+            hostsToRecommission.entrySet().removeIf(e -> disconnectedNMHostNames.contains(StringUtils.lowerCase(e.getKey())));
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleCommissionViaCMHandler.java:171]: Like above.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java (hunk 13 above)
Comment (Vatsal23082000) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java:189]: Here, The input will be static content and String.toLowerCase() is a non-static method. So, StringUtils.lowerCase() is used.

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 15]
```diff
@@ -124,14 +148,33 @@ public DetailedHostStatuses getDetailedHostStatuses(Optional<String> runtimeVers
...
+    private List<ApiRole> getNodemanagerRolesFromCM(StackDtoDelegate stack) {
+        RolesResourceApi rolesResourceApi = clouderaManagerApiFactory.getRolesResourceApi(apiClient);
+        try {
+            String yarnServiceName = extractYarnServiceNameFromBlueprint(stack);
+            ApiRoleList roles = rolesResourceApi.readRoles(stack.getName(), yarnServiceName, null, FULL_WITH_HEALTH_CHECK_EXPLANATION);
+            LOGGER.info("Fetched ApiRoleList from CM RolesResourceAPI readRoles call: {}", roles);
+            return emptyIfNull(roles.getItems()).stream().filter(role -> YarnRoles.NODEMANAGER.equalsIgnoreCase(role.getType())).toList();
+        } catch (ApiException e) {
+            LOGGER.error("Error when trying to read roles from CM", e);
             throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Done

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 16]
```diff
@@ -165,6 +208,26 @@ private static Optional<DetailedServicesHealthCheck> getDetailedServicesHealthCh
...
+                .map(nm -> nm.getHostRef().getHostname())
+                .map(StringUtils::lowerCase)
+                .collect(Collectors.toUnmodifiableSet());
+    }
+
+    private boolean isNodeManagerDisconnected(ApiRole role) {
+        Predicate<ApiHealthCheck> disconnectedNMCheck = hc -> {
+            ApiHealthSummary summary = hc.getSummary();
+            String healthCheckName = hc.getName();
+            // The connectivty health check is disabled if the nodes are deocmmissioned
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Done

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 17]
```diff
@@ -165,6 +208,26 @@ private static Optional<DetailedServicesHealthCheck> getDetailedServicesHealthCh
...
+            return NODE_MANAGER_CONNECTIVITY.equals(healthCheckName)
+                    && !DISABLED.equals(summary) && (BAD.equals(summary) || CONCERNING.equals(summary));
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Removed !DISABLED.equals(summary)

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java [hunk 18]
```diff
@@ -174,19 +237,23 @@ private static Optional<DetailedCertHealthCheck> getDetailedCertificateCheck(Api
...
+            case STOPPED, STOPPING -> true;
+            default -> false;
+        };
+    }
+
+    @VisibleForTesting
+    protected String extractYarnServiceNameFromBlueprint(StackDtoDelegate stack) {
+        CmTemplateProcessor templateProcessor = cmTemplateProcessorFactory.get(stack.getBlueprint().getBlueprintText());
+        return templateProcessor.getTemplate().getServices().stream().filter(service -> YarnRoles.YARN.equalsIgnoreCase(service.getServiceType()))
+                .map(ApiClusterTemplateService::getRefName).findAny().orElse(null);
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: changed

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (lajosrodek) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: The only change I see is the inclusion of a message:
```
throw new RuntimeException("Error when trying to read roles from CM", e);
```
Anyway, the exception type is not a blocker here. :smile:

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java (hunk 13 above)
Comment (lajosrodek) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java:189]: Let me ask differently:
```
e.getKey().toLowerCase()
```
vs
```
StringUtils.lowerCase(e.getKey())
```

As I see, `StringUtils.lowerCase()` is implemented this way:
```
    public static String lowerCase(final String str) {
        if (str == null) {
            return null;
        }
        return str.toLowerCase();
    }
```
So the only difference here is the latter being null-safe, whereas `e.getKey().toLowerCase()` will obviously throw an NPE if any key in `hostsToRemove` is `null`. Is this something we expect to be prepared for? :smile:

Comment (lajosrodek) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java:189]: Anyway, this is only a subtlety, so no need to change it.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleCommissionViaCMHandler.java (hunk 14 above)
Comment (lajosrodek) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleCommissionViaCMHandler.java:171]: Same as above, no need to change this now.

**Review by lajosrodek — APPROVED** [review]
Thanks, LGTM.

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (aakashshankar) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: An update on this. This is a crucial check to make sure fast upscale works. For hosts with the NodeManager role in STOPPED state, this heath check param is always DISABLED. Otherwise, it reports GOOD/BAD etc (depending on connectivity between NM and ResourceManager). A follow up to fix this has to be introduced

[the remark this answers is not in the document]
Comment (lajosrodek) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterHealthService.java]: Let `e1 = !DISABLED.equals(summary) && (BAD.equals(summary) || CONCERNING.equals(summary))` and `e2 = BAD.equals(summary) || CONCERNING.equals(summary)`. Then we can observe the following results depending on the value of `summary`:

* `summary == DISABLED`: `e1 == e2 == false`.
* `summary == BAD`: `e1 == e2 == true`.
* `summary == CONCERNING`: `e1 == e2 == true`.
* `summary` is some other value (so not `DISABLED`, `BAD` or `CONCERNING`): `e1 == e2 == false`.

Sorry but I do not see how the value of `summary == DISABLED` could influence the outcome here.
